specification/dartLangSpec.tex

```diff
@@ -13921,62 +13931,143 @@ \subsection{Exports}
 \LMHash{}%
-A library always exports all names and all declarations in its public namespace.
-In addition, a library may choose to re-export additional libraries via \Index{export directives}, often referred to simply as \Index{exports}.
+We need to combine two namespaces,
+which is an operation that is defined
+when they have disjoint sets of keys:
+The
+\IndexCustom{union of two namespaces}{namespace!union},
+\IndexCustom{$\Namespace{a}\cup\Namespace{b}$}{%
+  $\cup$@$\Namespace{a} \cup \Namespace{b}$},
+is the namespace that maps
+each key $n$ of \Namespace{a}
+to the corresponding value $\Namespace{a}(n)$,
+and each key $n$ of \Namespace{b}
+to $\Namespace{b}(n)$.
 
 \LMHash{}%
-Let $E$ be an export directive that refers to a URI via the string $s_1$.
-Evaluation of $E$ proceeds as follows:
+The exported namespace of a library is then determined as follows.
+Let $L$ be a library,
+let \List{E}{1}{m} be the export directives of $L$,
+and let \List{L}{1}{m} be libraries
+such that $E_i$ refers to $L_i$ for all $i \in 1 .. m$.
 
 \LMHash{}%
-First,
+Let \Namespace{L} be the namespace that
+for each public top-level declaration $D$ with name $n$ in $L$ maps $n$ to $D$.
+For each $i \in 1 .. m$,
+let $\Namespace{\metavar{export}, i}$ be the exported namespace of $L_i$,
+and let \Namespace{i} be a namespace obtained from
+$\Namespace{\metavar{export}, i}$ by the narrowing process described below.
 
-\begin{itemize}
-\item
-If the URI that is the value of $s_1$ has not yet been accessed by an import or export directive in the current isolate then the contents of the URI are compiled to yield a library $B$.
-\item Otherwise, the contents of the URI denoted by $s_1$ have been compiled into a library $B$ within the current isolate.
-\end{itemize}
+\LMHash{}%
+The narrowing process that yields
+\Namespace{i} from $\Namespace{\metavar{export}, i}$
+consists of two steps,
+each of them restricting the domain of the namespace to a subset
+(\commentary{%
+that is, deleting the binding from a key to a value,
+for certain keys%
+}).
+
+{ %% Local scope: abbreviate narrowing namespaces.
+
+% The sequence of narrowed name spaces of export i.
+\def\NSNi#1{\Namespace{\metavar{narrowing},i,{#1}}}
+
+% The final narrowed name space of a given export. We don't want to
+% mention the number of hide/show clauses when the export varies.
+\def\NSN#1{\Namespace{\metavar{narrowing},{#1}}}
 
 \LMHash{}%
-Let $NS_0$ be the exported namespace of $B$.
-Then, for each combinator clause $C_i, i \in 1 .. n$ in $E$:
+Let \NSNi{0} be \Namespace{\metavar{export}, i}.
+Then, for each combinator clause $C_j$, $j \in 1 .. l$ in $E_i$:
+
 \begin{itemize}
-\item If $C_i$ is of the form \code{\SHOW{} $\id_1, \ldots,\ \id_k$} then let
+\item If $C_j$ is of the form \code{\SHOW{} $\id_1, \ldots,\ \id_k$} then let
 
-$NS_i = \SHOW{}([\id_1, \ldots,\ \id_k], NS_{i-1}$).
+$\NSNi{j} = \SHOW{}([\id_1, \ldots,\ \id_k], \NSNi{j-1})$.
```

Comment:
Consider cross-referencing the "$hide(l,n)" and "$show(l,n)" meta-combinators defined in the import section?  Otherwise this bit is very opaque when read in isolation.

Reply:
Done.